orttraining/orttraining/training_ops/cuda/communication/send.cc

```diff
@@ -3,69 +3,119 @@
 
 #ifdef USE_HOROVOD
 
-#include "send.h"
+#include "orttraining/training_ops/cuda/communication/send.h"
+#include "orttraining/training_ops/cuda/communication/common.h"
+#include <mpi.h>
 
 namespace onnxruntime {
 namespace cuda {
 
-#define REGISTER_SEND_KERNEL_TYPED(T)                                    \
-  ONNX_OPERATOR_TYPED_KERNEL_EX(                                         \
-      Send,                                                              \
-      kMSDomain,                                                         \
-      1,                                                                 \
-      T,                                                                 \
-      kCudaExecutionProvider,                                            \
-      KernelDefBuilder()                                                 \
-          .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */   \
-          .OutputMemoryType<OrtMemTypeCPUOutput>(0) /* CPU variable */   \
-          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())         \
-          .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>()), \
-      Send<T>);
-
-REGISTER_SEND_KERNEL_TYPED(MLFloat16)
-REGISTER_SEND_KERNEL_TYPED(float)
-REGISTER_SEND_KERNEL_TYPED(double)
+ONNX_OPERATOR_KERNEL_EX(
+    Send,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */
+        .InputMemoryType<OrtMemTypeCPUInput>(1)   /* CPU variable */
+        .OutputMemoryType<OrtMemTypeCPUOutput>(0) /* CPU variable */
+        .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>())
+        .TypeConstraint("TInt64", DataTypeImpl::GetTensorType<int64_t>())
+        .TypeConstraint("V", DataTypeImpl::AllFixedSizeTensorTypes()),
+    Send);
 
 void CUDART_CB HostSend(void* args) {
   CommInfo_t* info = reinterpret_cast<CommInfo_t*>(args);
-  int mpi_code  = MPI_Send(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD);
+  int mpi_code = MPI_Send(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD);
   ORT_ENFORCE(mpi_code == MPI_SUCCESS, "MPI Send fails.");
 }
 
-template <typename T>
-Status Send<T>::ComputeInternal(OpKernelContext* ctx) const {
+Status Send::ComputeInternal(OpKernelContext* ctx) const {
   // Check if control signal is true.
   const Tensor* input_signal_tensor = ctx->Input<Tensor>(0);
   const bool* input_signal = input_signal_tensor->template Data<bool>();
   ORT_ENFORCE(*input_signal, "Input control signal of Send must be true before executing the node.");
```

Comment:
why we need this input? The signal should be handled through WaitEvent/RecordEvent. When executing send/recv, the order should already be guarded by wait/record.

Reply:
It is just from the spec. 
Bool is used determine the ort topology order.

Reply:
It seems this Boolean is always true, and the topology order can be determined through the variadic input. In that case, I think we should avoid sending data when it is not necessary.

Reply:
bool is true, when it runs. 
It is a safety check.
I was not talking about dependency of data tensor. I was talking about control signal.
Of course we can optimize it. 
From my understanding in the original doc, the spec is designed to support both single send/recv and aggregated send/recv, and many various pattern, like send, send and then WaitEvent.